### DISCUSSION TITLE
calculating tremor with tremor script

### DISCUSSION THREAD

**Discussion by DL854359743057** [YannickJadoul/Parselmouth#92] [Q&A]
Hi,
Thanks for developing this great tool!
I try to run the tremor script in parselmouth.
`tremor = parselmouth.praat.run_file(snd, 'tremor2.07/tremor.praat', "Analyis mode", 0.015, 60, 350, 0.03, 0.3, 0.01, 0.35, 0.14, "Envelope [To AmplitudeTier (period)]", 1.5, 15, 0.01, 0.15, 0.01, 0.01)`
However, I get the following error:
`parselmouth.PraatError: Cannot view or edit a Pitch from batch.`
How can I resolve this error?

**Comment by YannickJadoul** [comment]
Hi @DL854359743057. Thanks for the kinds words :-)
So yes, bad news is that it seems that this script is using the "View & Edit" editor action, which cannot be used in Praat in batch mode (i.e., when you run a Praat script directly from the terminal, rather than from the GUI). Now, as Parselmouth just provides the non-GUI functionality, it also uses batch mode, and these GUI functions cannot be used (just as if you'd run `praat tremor2.7/tremor.praat "Analysis mode" 0.015 ...` from the terminal).

The good news is that all of this functionality in "View & Edit" editor windows can typically also just be called from the main window, and so you would be able to make some minor modifications to the script to work with Parselmouth in Python. I've had people ask about this before, already, [on Gitter](https://gitter.im/PraatParselmouth/Lobby?at=5e6121d87fef7f2e89a24787). So I will tag @Rachine and @hadware, as they were dealing with this back then.

That being said, I'll also have a quick look. Am I correctly inferring that you are using [this script](https://www.researchgate.net/publication/317579089_tremorpraat_207)?

**Reply by YannickJadoul** [reply]
I found one hack'ish solution. The editor windows are only called opened during "Analysis mode". But we can create an extra mode:

Change the start of `tremor.praat` to look like this:
```
form Tremor 2.07
#   comment Select program mode
   optionmenu Mode 1
      option Analyis mode
      option Run mode
      option Parselmouth mode
```
and the end like this:
```
if mode = 1
   call anainout
elsif mode = 2
   call rinout
else
   call anainout
endif
```

If I now run this in Python:
```python
tremor, output = parselmouth.praat.run_file(sound, 'tremor2.07/tremor.praat', "Parselmouth mode", 0.015, 60, 350, 0.03, 0.3, 0.01, 0.35, 0.14, "Envelope [To AmplitudeTier (period)]", 1.5, 15, 0.01, 0.15, 0.01, 0.01, capture_output=True)
```
`output` will contain this string:
```
'Soundname: untitled\n\nfrequency contour magnitude (FCoM): 0.963\namplitude contour magnitude (ACoM): 0.939\n\nfrequency tremor cyclicality (FTrC): 0\namplitude tremor cyclicality (ATrC): 0.339\n\nfrequency tremor frequency (FTrF): --undefined-- Hz\namplitude tremor frequency (ATrF): 2.406 Hz\n\nfrequency tremor intensity index (FTrI): --undefined-- %\namplitude tremor intensity index (ATrI): 81.843 %\n\nfrequency tremor power index (FTrP): --undefined--\namplitude tremor power index (ATrP): 57.815\n'
```

Or, like this, if you print that string:
```
$ python test_tremor.py 
Soundname: untitled

frequency contour magnitude (FCoM): 0.963
amplitude contour magnitude (ACoM): 0.939

frequency tremor cyclicality (FTrC): 0
amplitude tremor cyclicality (ATrC): 0.339

frequency tremor frequency (FTrF): --undefined-- Hz
amplitude tremor frequency (ATrF): 2.406 Hz

frequency tremor intensity index (FTrI): --undefined-- %
amplitude tremor intensity index (ATrI): 81.843 %

frequency tremor power index (FTrP): --undefined--
amplitude tremor power index (ATrP): 57.815

```

**Reply by YannickJadoul** [reply]
If you don't want to parse that output, there's an even better/easier thing to do:

```python
tremor, _, variables = parselmouth.praat.run_file(sound, 'tremor2.07/tremor.praat', "Parselmouth mode", 0.015, 60, 350, 0.03, 0.3, 0.01, 0.35, 0.14, "Envelope [To AmplitudeTier (period)]", 1.5, 15, 0.01, 0.15, 0.01, 0.01, capture_output=True, return_variables=True)

for measure in ['ftrm', 'atrm', 'ftrc', 'atrc', 'ftrf', 'atrf', 'ftri', 'atri', 'ftrp', 'atrp']:
	print(f"{measure} = {variables[measure]}")
```

```
$ python test_tremor.py 
ftrm = 0.9634173132403819
atrm = 0.9391897756680819
ftrc = 0.0
atrc = 0.33884386080455386
ftrf = nan
atrf = 2.406109999445434
ftri = nan
atri = 81.84322534633904
ftrp = nan
atrp = 57.81486884591353
```

**Reply by YannickJadoul** [reply]
Finally, if you still want to use the "Run mode" of `tremor.praat`, you need to add two extra arguments to the form arguments. So adapt the final part of the `form`-`endform` block at the start of `tremor.praat` like this:

```
   positive Frequency_tremor_octave_cost 0.01
   positive Amplitude_tremor_octave_cost 0.01
   sentence path_of_sounds_to_be_analyzed ../
   sentence path_and_name_of_result_csv ../restab_tremor
endform
```

And then you can run the script like this:
```python
parselmouth.praat.run_file('tremor2.07/tremor.praat', "Run mode", 0.015, 60, 350, 0.03, 0.3, 0.01, 0.35, 0.14, "Envelope [To AmplitudeTier (period)]", 1.5, 15, 0.01, 0.15, 0.01, 0.01, "tmp_test_files/", "tremor_output", keep_cwd=True)
```

And `tremor_output.csv` will look like this:
```
soundname	FCoM	FTrC	FTrF [Hz]	FTrI [%]	FTrP	ACoM	ATrC	ATrF [Hz]	ATrI [%]	ATrP
the_north_wind_and_the_sun_10	0.963	0	--undefined--	--undefined--	--undefined--	0.939	0.339	2.406	81.843	57.815
```
